src/common/SubProcess.cc

```diff
@@ -397,20 +400,193 @@ void SubProcessTimed::exec() {
 }
 
 #else
+void SubProcess::close_h(HANDLE &handle) {
+  if (!handle)
+    return;
+
+  CloseHandle(handle);
+  handle = NULL;
+}
+
 int SubProcess::join() {
-  return EXIT_FAILURE;
+  ceph_assert(is_spawned());
+
+  close(stdin_pipe_out_fd);
+  close(stdout_pipe_in_fd);
+  close(stderr_pipe_in_fd);
+
+  DWORD status = 0;
+
+  if (WaitForSingleObject(proc_handle, INFINITE) != WAIT_FAILED) {
+    if (!GetExitCodeProcess(proc_handle, &status)) {
+      errstr << cmd << " : Could not get exit code: " << pid
+             << ". Error code: " << GetLastError() << "\n";
+    }
+  }
+  else {
+    errstr << cmd << " : Waiting for child process failed: " << pid
+           << ". Error code: " << GetLastError() << "\n";
+  }
+
+  if (status) {
+    errstr << cmd << ": exit status: " << status;
+  }
+
+  close_h(proc_handle);
+  pid = -1;
+
+  return status;
 }
 
 void SubProcess::kill(int signo) const {
+  ceph_assert(is_spawned());
+  ceph_assert(TerminateProcess(proc_handle, 128 + SIGTERM));
 }
 
 int SubProcess::spawn() {
-  return EXIT_FAILURE;
+  std::ostringstream cmdline;
+  cmdline << cmd;
+  for (std::vector<std::string>::iterator i = cmd_args.begin();
+       i != cmd_args.end();
+       i++) {
+    cmdline << " \"" << i->c_str() << "\"";
+  }
+
+  STARTUPINFO si = {0};
+  PROCESS_INFORMATION pi = {0};
+  SECURITY_ATTRIBUTES sa = {0};
+
+  sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+  sa.bInheritHandle = TRUE;
+  sa.lpSecurityDescriptor = NULL;
+
+  HANDLE stdin_r = INVALID_HANDLE_VALUE, stdin_w = INVALID_HANDLE_VALUE,
+         stdout_r = INVALID_HANDLE_VALUE, stdout_w = INVALID_HANDLE_VALUE,
+         stderr_r = INVALID_HANDLE_VALUE, stderr_w = INVALID_HANDLE_VALUE;
+
+  if ((stdin_op == PIPE && !CreatePipe(&stdin_r, &stdin_w, &sa, 0)) ||
```

Comment:
nit, could define a helper for creating pipe and error handling according to different `std_fd_op` instead of repeating the same logic 3 times everywhere.

Reply:
I'd rather keep it like this. I did try that approach at some point and it ended up being much harder to read than it is now.